qualtran/bloqs/arithmetic/addition.py

```diff
@@ -213,6 +216,120 @@ def __pow__(self, power: int):
         raise NotImplementedError("OutOfPlaceAdder.__pow__ defined only for +1/-1.")
 
 
+@frozen
+class SimpleAddConstant(Bloq):
+    r"""Applies U_{k}|x> = |x + k>.
+
+    Applies addition to input register `|x>` given classical parameter 'k'.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+        k: The classical value to be added to x.
+        cvs: A tuple of control variable settings. Each entry specifies whether that
+            control line is a "positive" control (`cv[i]=1`) or a "negative" control `0`.
+        signed: A boolean condition which controls whether the x register holds a value represented
+            in 2's Complement or Unsigned. This effects the ability to add a negative constant.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+
+    References:
+        [Improved quantum circuits for elliptic curve discrete logarithms](https://arxiv.org/abs/2001.09580) Fig 2a
+    """
+
+    bitsize: int
+    k: int
+    cvs: Tuple[int, ...] = field(converter=lambda v: (v,) if isinstance(v, int) else tuple(v))
+    signed: bool
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        if len(self.cvs) > 0:
+            return Signature(
+                [Register('ctrl', bitsize=len(self.cvs)), Register('x', bitsize=self.bitsize)]
+            )
+        else:
+            return Signature([Register('x', bitsize=self.bitsize)])
+
+    def binary_rep(self, num, num_bits):
+        # If the registers are interpreting signed bits, represent the output in 2's complement.
+        # Otherwise make sure the classical value k is > 0.
+        if self.signed:
+            num &= (2 << num_bits - 1) - 1
+        else:
+            assert self.k >= 0
+        format_str = '{:0' + str(num_bits) + 'b}'
+        return format_str.format(int(num))
+
+    def on_classical_vals(
+        self, ctrl: 'ClassicalValT', x: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        if ctrl == 0:
+            return {'ctrl': 0, 'x': x}
+
+        assert ctrl == 1, 'Bad ctrl value.'
+        x_out = x + self.k
+        return {'ctrl': ctrl, 'x': x_out}
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **regs: SoquetT) -> Dict[str, 'SoquetT']:
+
+        # Assign registers to variables and allocate ancilla bits for classical integer k.
```

Comment:
can you include a description of how this construction works in the class docstring